test/videojs-contrib-hls.test.js

```diff
@@ -1096,6 +1096,121 @@ QUnit.test('playlist 404 should blacklist media', function(assert) {
   assert.equal(this.player.tech_.hls.stats.bandwidth, 1e10, 'bandwidth set above');
 });
 
+QUnit.test('detect if the playlist stuck at the end', function(assert) {
+  let playlist;
+
+  this.player.src({
+    src: 'manifest/master.m3u8',
+    type: 'application/vnd.apple.mpegurl'
+  });
+  openMediaSource(this.player, this.clock);
+  this.standardXHRResponse(this.requests.shift());
+  let hls = this.player.tech_.hls;
+
+  playlist = hls.selectPlaylist();
+  // not stuck at playlist end when no seekable, even if empty buffer
+  // and positive currentTime
+  hls.masterPlaylistController_.seekable = function() {
+    return videojs.createTimeRange();
+  };
+  this.player.tech_.buffered = function() {
+    return videojs.createTimeRange();
+  };
+  this.player.tech_.setCurrentTime(170);
+  assert.ok(!hls.masterPlaylistController_.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
+
+  // not stuck at playlist end when no seekable, even if empty buffer
+  // and currentTime 0
+  this.player.tech_.setCurrentTime(0);
+  assert.ok(!hls.masterPlaylistController_.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
+
+  // not stuck at playlist end when no seekable but more buffer
+  this.player.tech_.buffered = function() {
+    return videojs.createTimeRange(0, 170);
+  };
+  assert.ok(!hls.masterPlaylistController_.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
+
+  // not stuck at playlist end when currentTime not at seekable end
+  // even if the buffer is empty
+  hls.masterPlaylistController_.seekable = function() {
+    return videojs.createTimeRange(0, 130);
+  };
+  this.player.tech_.setCurrentTime(50);
+  this.player.tech_.buffered = function() {
+    return videojs.createTimeRange();
+  };
+  Hls.Playlist.playlistEnd = function() {
+    return 130;
+  };
+  assert.ok(!hls.masterPlaylistController_.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
+
+  // stuck at playlist end when there is no buffer and playhead
+  // reached absolute end of playlist
+  this.player.tech_.setCurrentTime(160);
+  Hls.Playlist.playlistEnd = function() {
+    return 160;
+  };
+  assert.ok(hls.masterPlaylistController_.stuckAtPlaylistEnd_(playlist), 'stuck at playlist end');
+
+  // stuck at playlist end when current time reach to the buffer end
+  // and buffer has reached absolute end of playlist
+  hls.masterPlaylistController_.seekable = function() {
+    return videojs.createTimeRange(90, 130);
+  };
+  this.player.tech_.buffered = function() {
+    return videojs.createTimeRange(0, 170);
+  };
+  this.player.tech_.setCurrentTime(170);
+  Hls.Playlist.playlistEnd = function() {
+    return 170;
+  };
+  assert.ok(hls.masterPlaylistController_.stuckAtPlaylistEnd_(playlist), 'stuck at playlist end');
+});
```

Comment:
I think there might be one more interesting case, where seekable is a normal range, the buffer reaches the absolute end of the playlist, and current time is in the buffered range (but not yet at the absolute end, say 1 second before it). This case should return that we aren't yet stuck.